src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -53,54 +43,52 @@ internal sealed partial class SerializableOptionSet : OptionSet
+        /// only so we can shortcircuit <see cref="UnionWithLanguages"/>.
+        /// </summary>
+        private readonly Lazy<ImmutableHashSet<string>> _languages;
+
         private SerializableOptionSet(
-            ImmutableHashSet<string> languages,
             WorkspaceOptionSet workspaceOptionSet,
-            ImmutableHashSet<IOption> serializableOptions,
             ImmutableDictionary<OptionKey, object?> values,
             ImmutableHashSet<OptionKey> changedOptionKeysSerializable,
             ImmutableHashSet<OptionKey> changedOptionKeysNonSerializable)
         {
-            Debug.Assert(languages.All(RemoteSupportedLanguages.IsSupported));
-
-            _languages = languages;
             _workspaceOptionSet = workspaceOptionSet;
-            _serializableOptions = serializableOptions;
             _serializableOptionValues = values;
             _changedOptionKeysSerializable = changedOptionKeysSerializable;
             _changedOptionKeysNonSerializable = changedOptionKeysNonSerializable;
 
             Debug.Assert(values.Keys.All(ShouldSerialize));
             Debug.Assert(changedOptionKeysSerializable.All(optionKey => ShouldSerialize(optionKey)));
             Debug.Assert(changedOptionKeysNonSerializable.All(optionKey => !ShouldSerialize(optionKey)));
+
+            _languages = new Lazy<ImmutableHashSet<string>>(() => this.GetLanguagesAndValuesToSerialize().languages);
         }
 
         internal SerializableOptionSet(
-            ImmutableHashSet<string> languages,
             IOptionService optionService,
-            ImmutableHashSet<IOption> serializableOptions,
             ImmutableDictionary<OptionKey, object?> values,
             ImmutableHashSet<OptionKey> changedOptionKeysSerializable)
-            : this(languages, new WorkspaceOptionSet(optionService), serializableOptions, values, changedOptionKeysSerializable, changedOptionKeysNonSerializable: ImmutableHashSet<OptionKey>.Empty)
+            : this(new WorkspaceOptionSet(optionService), values, changedOptionKeysSerializable, changedOptionKeysNonSerializable: ImmutableHashSet<OptionKey>.Empty)
         {
         }
 
-        public ImmutableHashSet<string> Languages => _languages;
-
         /// <summary>
         /// Returns an option set with all the serializable option values prefetched for given <paramref name="languages"/>,
         /// while also retaining all the explicitly changed option values in this option set for any language.
-        /// NOTE: All the provided <paramref name="languages"/> must be <see cref="RemoteSupportedLanguages.IsSupported(string)"/>.
+        /// Note: All the provided <paramref name="languages"/> must be <see cref="RemoteSupportedLanguages.IsSupported(string)"/>.
         /// </summary>
-        public SerializableOptionSet WithLanguages(ImmutableHashSet<string> languages)
+        public SerializableOptionSet UnionWithLanguages(ImmutableHashSet<string> languages)
```

Comment:
So we are now basically tracking set of all languages over the lifetime of the solution rather than in the current solution snapshot, correct?

Reply:
correct. it is effectively just data.  and once we load options for a particular language, they are loaded and are in the snapshot.